File: 이것이 취업을 위한 코딩 테스트다/다이나믹_프로그래밍/효율적인_화폐_구성/효율적인_화폐_구성.cpp
#include <iostream>
#include <vector>
using namespace std;

int main() {
  int n, m, num;
  cin >> n >> m;
  vector<int> arr;
  vector<int> dp(m + 1, 10001);
  dp[0] = 0;

  for (int i = 0; i < n; i++) {
    cin >> num;
    arr.push_back(num);
  }

  for (int i = 0; i < n; i++) {
    int count = arr[i];

    for (int j = count; j <= m; j++) {
      dp[j] = min(dp[j], dp[j - count] + 1);
    }
  }

  if (dp[m] == 10001) {
    cout << -1;
  } else {
    cout << dp[m];
  }

  return 0;
}

File: 이것이 취업을 위한 코딩 테스트다/다이나믹_프로그래밍/효율적인_화폐_구성/효율적인_화폐_구성.java
import java.util.Arrays;
import java.util.Scanner;

public class 효율적인_화폐_구성 {
  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);

    int n = sc.nextInt();
    int m = sc.nextInt();

    int[] arr = new int[n];
    int[] dp = new int[m + 1];
    
    Arrays.fill(dp, 10001);
    dp[0] = 0;

    for (int i = 0; i < n; i++) {
      arr[i] = sc.nextInt();
    }

    for (int num: arr) {
      for (int i = num; i <= m; i++) {
        dp[i] = Math.min(dp[i], dp[i - num] + 1);
      }
    }

    if (dp[m] == 10001) {
      System.out.println(-1);
    } else {
      System.out.println(dp[m]);
    }
  }
}
